### ISSUE TITLE
APIChangesetNotSubscribedError has HTTP status code 404

### ISSUE THREAD

**Issue by westnordost** [openstreetmap/openstreetmap-website#1199]
**First off, I am not sure if this is a bug or intended behavior**:

In the changeset API, if a user tries to subscribe to a changeset he is already subscribed to, he gets a `APIChangesetAlreadySubscribedError` error. The other way round, if he is not subscribed and tries to unsubscribe, he gets a `APIChangesetNotSubscribedError`.

So far so good, but the HTTP status code for APIChangesetAlreadySubscribedError is a 409 - Conflict while the status code for APIChangesetNotSubscribedError is a 404 - Not Found. This seems to be inconsistent.

404 Not Found is also returned if the changeset one wants to subscribe/unsubscribe to/from does not exist at all. This inconsistency makes it harder for clients to report to the user what is the exact problem.

So... is this a bug?

**Comment by westnordost** [comment]
@ukasiu authored the addition of the changeset discussion. He should be able to answer this.

**Comment by mmd-osm** [comment]
> HTTP status code for APIChangesetAlreadySubscribedError is a 409 - Conflict while the status code for APIChangesetNotSubscribedError is a 404 - Not Found. This seems to be inconsistent.



I'd say this works as documented: https://wiki.openstreetmap.org/wiki/API_v0.6#Subscribe:_POST_.2Fapi.2F0.6.2Fchangeset.2F.23id.2Fsubscribe

> 404 Not Found is also returned if the changeset one wants to subscribe/unsubscribe to/from does not exist at all. This inconsistency makes it harder for clients to report to the user what is the exact problem.

If you're not sure if a changeset exists at all, you could use another API call to rule out non-existing changesets: https://wiki.openstreetmap.org/wiki/API_v0.6#Read:_GET_.2Fapi.2F0.6.2Fchangeset.2F.23id.3Finclude_discussion.3Dtrue

**Comment by gravitystorm** [comment]
> I'd say this works as documented: https://wiki.openstreetmap.org/wiki/API_v0.6#Subscribe:_POST_.2Fapi.2F0.6.2Fchangeset.2F.23id.2Fsubscribe

That's not right - the wiki documentation is not a specification, it's just documenting what currently happens. Therefore you can't use it to say that the code "works as documented".

**Comment by mmd-osm** [comment]
Although the Rails code will be the authoritive source for the actual behavior, you can't expect API consumers to study the Rails code to find out how the API is supposed to work.

It's very unfortunate that the Wiki isn't seen as reference and kept up to date by everyone accordingly.

My wish list for 2020: we really need written documentation that reflects current actual behavior.
